esp32/uart: Preserve console UART clock, fix UART console with DFS.

xyzzy42 opened the pull request:
In commit 7c929d44 the console UART was changed to use the UART HAL. Starting the UART HAL will change the UART clock from whatever it was already configured at to UART_SCLK_DEFAULT.  There is no "initialize at existing settings" option.

This clock doesn't work with DFS.

The ESP-IDF code already takes this into account, and when DFS is enabled it will configure the console UART to use the correct platform-specific clock that will work with DFS.

The UART HAL init undoes this and sets it back to default.

This change will query the clock before the HAL init, then use the HAL function to restore it back.  Thus keeping the clock at the "correct" value, which depends on platform, DFS status, and so on.

The clock frequency will be found using the UART driver function ESP-IDF code uses for this.  The existing code hard-coded a path that worked if the clock was the APB clock and would fail otherwise.

The UART_NUM_0 define is removed because driver/uart.h already provides this same macro.

xyzzy42 committed:
esp32/uart: Preserve console UART clock, fix UART console with DFS.

In commit 7c929d44 the console UART was changed to use the UART HAL.
Starting the UART HAL will change the UART clock from whatever it was
already configured at to UART_SCLK_DEFAULT.  There is no "initialize at
existing settings" option.

This clock doesn't work with DFS.

The ESP-IDF code already takes this into account, and when DFS is enabled
it will configure the console UART to use the correct platform-specific
clock that will work with DFS.

The UART HAL init undoes this and sets it back to default.

This change will query the clock before the HAL init, then use the HAL
function to restore it back.  Thus keeping the clock at the "correct"
value, which depends on platform, DFS status, and so on.

The clock frequency will be found using the UART driver function ESP-IDF
code uses for this.  The existing code hard-coded a path that worked if the
clock was the APB clock and would fail otherwise.

The UART_NUM_0 define is removed because driver/uart.h already provides
this same macro.

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

dpgeorge commented:
Thanks for the contribution.

It looks OK to me.  It does introduce a dependency on the `driver/uart.h` API but I guess that's necessary to compute the clock frequency?

@projectgus any comments on this?

xyzzy42 commented:
One could duplicate the code to compute it, or only support some sub-set of possible clocks.  E.g., the previous code only supported APB clock.  Both seem very hard to maintain.

The driver/uart.h API is used by the machine_uart.c driver, so there is likely already going to be another user of the API.

projectgus commented on ports/esp32/uart.c:
```diff
@@ -53,16 +50,14 @@ STATIC void uart_irq_handler(void *arg);
+    ESP_ERROR_CHECK(uart_get_sclk_freq(sclk, &sclk_freq));
 
     uart_hal_init(&repl_hal, MICROPY_HW_UART_REPL); // Sets defaults: 8n1, no flow control
+    uart_hal_set_sclk(&repl_hal, sclk);
```
```suggestion
    uart_hal_set_sclk(&repl_hal, sclk); // Restore the previously configured SCLK
```

projectgus commented on ports/esp32/uart.c:
```diff
@@ -34,12 +34,9 @@
 #if MICROPY_HW_ENABLE_UART_REPL
 
 #include <stdio.h>
+#include "driver/uart.h"
```
```suggestion
#include "driver/uart.h" // For uart_get_sclk_freq()
```

projectgus reviewed (approved):
Thanks for the detailed description of the problem and the fix, @xyzzy42 .

The original intention of switching to HAL in 7c929d44 was to remove the dependency on the driver layer entirely, but I see that it's going to be fiddly to otherwise recreate the function `uart_get_sclk_freq()` here for the different possible clock sources (arguably that function should be an IDF HAL function, but that's not really our argument to make!)

I've made a couple of suggestions for comments to make it clearer to a future reader who hasn't seen this PR why this code is included like this, but otherwise it looks very good to me.

projectgus commented on ports/esp32/uart.c:
```diff
@@ -53,16 +50,14 @@ STATIC void uart_irq_handler(void *arg);
-    // ESP32-H2 and ESP32-C2, I think
-    #error "This SoC uses a different default UART SCLK source, code needs updating."
-    #endif
+    uart_hal_get_sclk(&repl_hal, &sclk);
```
```suggestion
    uart_hal_get_sclk(&repl_hal, &sclk); // Avoid resetting SCLK in uart_hal_init()
```

xyzzy42 committed:
esp32/uart: Make compatible with sclk type change in ESP-IDF 5.3.

This type changes when they add LP-UART support for the ESP32-C6 in the dev
version of ESP-IDF 5.3.

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

xyzzy42 commented:
> (arguably that function should be an IDF HAL function, but that's not really our argument to make!)

It doesn't interact with the driver layer abstraction of the UART in any way, so it would really fit better at the HAL layer.  I wonder if `esp_clk_tree_src_get_freq_hz()` is considered a driver layer function, and they don't want a HAL function to call a driver function in another driver?

xyzzy42 commented:
I've added a patch to be compatible with the current version of ESP-IDF.
